testutil/fuzz.go

```diff
@@ -0,0 +1,114 @@
+// Copyright © 2022-2023 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package testutil
+
+import (
+	"testing"
+
+	v1deneb "github.com/attestantio/go-eth2-client/api/v1/deneb"
+	eth2spec "github.com/attestantio/go-eth2-client/spec"
+	"github.com/attestantio/go-eth2-client/spec/altair"
+	"github.com/attestantio/go-eth2-client/spec/deneb"
+	eth2p0 "github.com/attestantio/go-eth2-client/spec/phase0"
+	fuzz "github.com/google/gofuzz"
+	"github.com/prysmaticlabs/go-bitfield"
+
+	"github.com/obolnetwork/charon/core"
+)
+
+// NewEth2Fuzzer returns a fuzzer for valid eth2 types.
+//
+// Note go-eth2-client Versioned*Blocks are not support, instead use core.Versioned*Blocks.
+func NewEth2Fuzzer(t *testing.T) *fuzz.Fuzzer {
+	t.Helper()
+
+	blindedVersions := []eth2spec.DataVersion{
+		eth2spec.DataVersionBellatrix,
+		eth2spec.DataVersionCapella,
+		// eth2spec.DataVersionDeneb,
+	}
+
+	allVersions := []eth2spec.DataVersion{
+		eth2spec.DataVersionPhase0,
+		eth2spec.DataVersionAltair,
+		eth2spec.DataVersionBellatrix,
+		eth2spec.DataVersionCapella,
+		eth2spec.DataVersionDeneb,
+	}
+
+	return fuzz.New().
+		NilChance(0).
+		Funcs(
+			// bitfield.Bitlist trailing byte must not be zero
+			func(e *bitfield.Bitlist, c fuzz.Continue) {
+				c.FuzzNoCustom(e)
+				l := len(*e)
+				(*e)[l-1] = 1
+			},
+			// eth2p0.AttesterSlashings may not be more than 2
+			func(e *[]*eth2p0.AttesterSlashing, c fuzz.Continue) {
+				c.FuzzNoCustom(e)
+				if len(*e) > 2 {
+					*e = (*e)[:2]
+				}
+			},
+			// Eth1Data.BlockHash must be 32 bytes
+			func(e *eth2p0.ETH1Data, c fuzz.Continue) {
+				c.FuzzNoCustom(e)
+				e.BlockHash = RandomBytes32()
+			},
+			// Just zero BeaconBlockBody.Deposits to pass validation.
+			func(e *[]*eth2p0.Deposit, c fuzz.Continue) {
+				*e = []*eth2p0.Deposit{}
+			},
+			// SyncAggregate.SyncCommitteeBits must have 64 bits
+			func(e *altair.SyncAggregate, c fuzz.Continue) {
+				c.FuzzNoCustom(e)
+				bits := bitfield.NewBitvector512()
+				for i := 0; i < 64; i++ {
+					bits.SetBitAt(uint64(i), true)
+				}
+				e.SyncCommitteeBits = bits
+			}, // SyncCommitteeContribution.AggregationBits must have 16 bits
+			func(e *altair.SyncCommitteeContribution, c fuzz.Continue) {
+				c.FuzzNoCustom(e)
+				bits := bitfield.NewBitvector128()
+				for i := 0; i < 16; i++ {
+					bits.SetBitAt(uint64(i), true)
+				}
+				e.AggregationBits = bits
+			},
+			// VersionBlingedBeaconBlock.Deneb SSZ not supported by goeth2client yet, so nil it.
+			func(e **v1deneb.BlindedBeaconBlock, c fuzz.Continue) {
+				*e = nil
+			},
+			// []deneb.KzgCommitment has max size 4.
```

Comment:
what if these bounds change in the future, for ex, `len(deneb.KZGCommitment)` increased to 16. Suggest using constants for these. This would also ensure stuff in the godoc getting out of sync

Reply:
can do cleanups later

Reply:
sure